drivers/disk/flashdisk.c

```diff
@@ -52,38 +55,66 @@ static int disk_flash_access_status(struct disk_info *disk)
 
 static int disk_flash_access_init(struct disk_info *disk)
 {
-	if (flash_dev) {
-		return 0;
+	struct mtd_disk_data *ctx = NULL;
+	int rc;
+	struct flash_pages_info page;
+
+	if (disk) {
+		ctx = CONTAINER_OF(disk, struct mtd_disk_data, info);
 	}
 
-	flash_dev = device_get_binding(CONFIG_DISK_FLASH_DEV_NAME);
-	if (!flash_dev) {
+	if (disk == NULL || ctx == NULL) {
+		LOG_ERR("Failed to obtain context pointer");
 		return -ENODEV;
 	}
 
+	if (!device_is_ready(disk->dev)) {
+		LOG_ERR("Device is not ready");
+		return -ENODEV;
+	}
+
+	rc = flash_get_page_info_by_offs(disk->dev, ctx->offset, &page);
+	if (rc != 0) {
+		LOG_ERR("Error %d while getting page info", rc);
+		return rc;
+	}
+
+	ctx->write_bs = flash_get_write_block_size(disk->dev);
+	ctx->page_size = page.size;
+	LOG_INF("Initialize device %s", disk->name);
+	LOG_INF("offset %ld, write bs %zu, page size %zu, volume size %zu",
+		ctx->offset, ctx->write_bs, ctx->page_size, ctx->size);
+
 	return 0;
 }
 
 static int disk_flash_access_read(struct disk_info *disk, uint8_t *buff,
 				uint32_t start_sector, uint32_t sector_count)
 {
+	struct mtd_disk_data *ctx;
 	off_t fl_addr;
 	uint32_t remaining;
 	uint32_t len;
 	uint32_t num_read;
 
-	fl_addr = lba_to_address(start_sector);
-	remaining = (sector_count * SECTOR_SIZE);
-	len = CONFIG_DISK_FLASH_MAX_RW_SIZE;
+	ctx = CONTAINER_OF(disk, struct mtd_disk_data, info);
+
+	fl_addr = ctx->offset + start_sector * SECTOR_SIZE;
+	if (fl_addr >= (ctx->offset + ctx->size)) {
+		LOG_ERR("Address %lx outside partition boundary", fl_addr);
+		return -EINVAL;
+	}
 
-	num_read = GET_NUM_BLOCK(remaining, CONFIG_DISK_FLASH_MAX_RW_SIZE);
+	remaining = (sector_count * SECTOR_SIZE);
+	len = ctx->page_size;
```

Comment:
> `flash_read()` call is able to read any length of data.
> Are you aware of any implementation restriction? (I'm not)

Neither am I :-) , not yet. I have not changed any logic, but one can improve almost every flash API related line here, I guess.  I think steam_flash could also be moved here or used here. In any case, it would be nice to have primitive caching mechanism here.

Reply:
steam_flash is designed to write block by block. I think not really useful here.